src/module.c

```diff
@@ -8614,8 +8614,14 @@ void firePostExecutionUnitJobs(void) {
  * infinite loops by halting the execution could result in violation of the feature correctness
  * and so Redis will make no attempt to protect the module from infinite loops.
  *
- * 'free_pd' can be NULL and in such case will not be used. */
+ * 'free_pd' can be NULL and in such case will not be used.
+ *
+ * Return REDISMODULE_OK on success and REDISMODULE_ERR if was called while loading data from disk (AOF or RDB) or
+ * if the instance is a readonly replica. */
 int RM_AddPostNotificationJob(RedisModuleCtx *ctx, RedisModulePostNotificationJobFunc callback, void *privdata, void (*free_privdata)(void*)) {
+    if (server.loading|| (server.masterhost && server.repl_slave_ro)) {
```

Comment:
trying to challenge this to be on the safe side:
what if the module wants to react to changes, and instead of propagating the reaction, it wants to perform a similar reaction in the replica?

e.g. on every key that's added, increment some counter.

well, if the counter is in the module's global memory (not part of the redis data set), then it can do that from the KSN, no need for PEJ.
but what if the counter is a key? it did have a PEJ in the master to increment that key, and it could have propagated that increment. but maybe for something more complicated than a counter, propagation is excessive and the module would rather compute it locally and modify the keyspace?

in theory, maybe we can match the client in the CTX to the master client, like what's done here:
```
    if (server.masterhost && server.repl_slave_ro &&
        !obey_client &&
        is_write_command)
```
calling @guybe7 and @yossigo to comment too.

Reply:
@oranagra I agree with you about it, we need some good arguments about why no PEJ use case is applicable in these cases.

Reply:
> in theory, maybe we can match the client in the CTX to the master client, like what's done here:

Then we also need to support PEJ on AOF because this logic effect will not be propagated to it. If the propagation is expensive then the module could have register a command that perform this logic and execute this command on the PEJ and replicate it as well. Complex propagation is not a good argument here in my opinion because it can be avoided easily.

Reply:
discussed it with @guybe7.
initially i argued that if a module wants to implement something similar to a script, and do script replication (rather than effects replication), it'll need to do the same modifications it did on the master, on the replica as well.
but that's actually about module that implements commands, not one that listens on KSN of other commands.

so for KSN the reasoning is that module can listen on them to update it's global state (index / counters), and that's not considered a write from redis's point of view, but if it decides to write that data into a key (inside the PEJ), it should propagate that, and then it doesn't need, and shouldn't do the same on the replica / AOF loading.

not sure i managed to convey the entire discussion, @guybe7 feel free to elaborate, but the bottom line is that we agree with the current PR that wishes to block this.

Reply:
@yossigo i'd like your final ok for merging this PR as is..